Test/comp/Arrays.dfy

```diff
@@ -374,3 +383,494 @@ module TypeSynonym {
+
+    var m0 := map[4 := true, 3 := false];
+    var m1 := map[4 := true][3 := false];
+    var m2 := map[3 := false][4 := true][3 := false][4 := true];
+    var m3 := map[3 := false][4 := false][3 := false][3 := false];
+    CheckEquality(m0, m1); // true
+    CheckEquality(m0, m2); // true
+    CheckEquality(m0, m3); // false
+  }
+
+  method StringToByteArray(s: string) returns (arr: array<byte>) {
+    arr := new [|s|];
+    forall i | 0 <= i < |s| {
+      arr[i] := (var ch := s[i]; if ' ' <= ch <= 'z' then ch else 'X') as byte;
+    }
+  }
+
+  method WriteLine(arr: array<byte>) {
+    for i := 0 to arr.Length {
+      print [arr[i] as char];
+    }
+    print "\n";
+  }
+
+  method CheckEquality<T(==)>(x: T, y: T) {
+    print " ", x == y, "\n";
+  }
+}
+
+module NativeArrays {
+  newtype byte = x | 0 <= x < 256
+  newtype onebyte = x | 0 < x < 256 witness 1
+
+  method Test() {
+    var aa := new byte[3];
+    PrintArray(aa, "uninitialized bytes");
+    aa := new byte[] [7, 8, 9];
+    PrintArray(aa, "bytes from display");
+    aa := new byte[3](i => if 0 <= i < 10 then (20 + i) as byte else 88);
+    PrintArray(aa, "bytes from lambda");
+
+    var bb := new onebyte[3];
+    PrintArray(bb, "uninitialized 1bytes");
+    bb := new onebyte[] [7, 8, 9];
+    PrintArray(bb, "1bytes from display");
+    bb := new onebyte[3](i => if 0 <= i < 10 then (20 + i) as onebyte else 88);
+    PrintArray(bb, "1bytes from lambda");
+
+    TestVariousFlavorsOfIndices();
+  }
+
+  method PrintArray(a: array, title: string) {
+    print title, ": array[";
+    for i := 0 to a.Length {
+      print if i == 0 then "" else ", ", a[i];
+    }
+    print "]\n";
+  }
+
+  method TestVariousFlavorsOfIndices() {
+    var arr := new int[250];
+    var m := new int[20, 30];
+
+    arr[3 as bv19] := 17;
+    arr[4 as bv5] := 19;
+    var iIndex: int, bIndex: byte, bvIndex: bv9 := 3, 4, 5;
+    m[iIndex, bIndex] := arr[iIndex];
+    arr[iIndex] := m[iIndex, bvIndex - 1];
+    assert arr[iIndex] == 17;
+    print arr[iIndex], " "; // 17
+    m[bIndex, iIndex] := arr[bIndex];
+    arr[bIndex] := m[bvIndex - 1, iIndex];
+    assert arr[bIndex] == 19;
+    print arr[bIndex], " "; // 19
+    m[iIndex, iIndex] := arr[iIndex] + 1;
+    arr[iIndex] := m[iIndex, iIndex];
+    m[bIndex, bIndex] := arr[bIndex] + 1;
+    arr[bIndex] := m[bIndex, bIndex];
+    print arr[iIndex], " ", arr[bIndex], "\n"; // 18 20
+  }
+}
+
+module SimultaneousAssignment {
+  method Test() {
+    var arr := new int[250];
+    var m := new int[20, 30];
+    var arr', m' := arr, m;
+
+    var a, b, c := 18, 28, 38;
+    var arr'' := new int[1];
+    var m'' := new int[1, 1];
+
+    arr, a, arr[a + b], m, arr, c, m[a, b], b, m := arr'', 100, 120, m'', arr'', 300, c, 200, m'';
+
+    print a, " ", b, " ", c, " ", arr'[46], " ", m'[18, 28], "\n"; // 100 120 200 300 38
+  }
+}
+
+module ArrayToSeq {
+  newtype byte = x | 0 <= x < 256
+
+  trait TraitMeRite {
```

Comment:
Open your eyes, maybe you'll see the light... :)

Reply:
I hit it with my best shot.